### PR TITLE
Pass thru more config vars

### PR DESCRIPTION
**Pull request by tzaffi**
# Allowing Greater Control Over Sandbox's Configuration

In swapping out the old docker build with a Sandbox based build, some minor tweaks were to the configs were necessitated (see the [SDK Testing PR](https://github.com/algorand/algorand-sdk-testing/pull/220) along with an example [Python SDK PR](https://github.com/algorand/py-algorand-sdk/pull/367) for how these tweaks are utilized)

## Summary of Changes

* When `.env` is present in a docker build's context, it is used to define environment variables. I'm introducing the ability to so with:
  * an empty NOOP `.env`
  * un-ignore it in the `.dockerignore`
  * source it in the `sandbox` script (so I'm proposing to use this as a "double duty" - both for configuring the docker build as well as the `./sandbox` script)

* New environment variables being passed through.
  * `NODE_ARCHIVAL` - when set `True` it will ensure that `algod`'s node is archival. This is achieved by passing the `--archival` flag to `images/algod/Dockerfile`'s `setup.py` script which sets the node's config variable `Archival=true`
  * `NODE_V1_INDEXER` - when set `True` it will ensure that `algod`'s node is brought up with a v1 indexer. This is achieved by passing the `--is_indexer_active` flag to `images/algod/Dockerfile`'s `setup.py` script which sets the node's config variable `isIndexerActive=true`
* Previously, the exposed docker ports were hard coded. Now allow these to be set via the `.env`. In particular, one can specify:
  * `ALGOD_PORT`
  * `KMD_PORT`
  * `CDT_PORT`
  * `INDEXER_PORT`
  * `POSTGRES_PORT`
* Previously, the docker container names were hard coded. Now allow these to be set via the `.env`. You can specify the names with:
  * `ALGOD_CONTAINER`
  * `INDEXER_CONTAINER`
  * `POSTGRES_CONTAINER`

## Testing
* Passes CI for this repo
* Passes Python SDK's CI.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -20,12 +20,15 @@ services:
...
     ports:
-      - 4001:4001
-      - 4002:4002
-      - 9392:9392
+      - ${ALGOD_PORT:-4001}:4001
+      - ${KMD_PORT:-4002}:4002
+      - ${CDT_PORT:-9392}:9392
+    privileged: true
+    volumes:
+      - ./volumes/algod:/opt/mnt
```
Comment (tzaffi) [inline docker-compose.yml:28]:
```suggestion
```

File: docker-compose.yml [hunk 2]
```diff
@@ -44,16 +47,20 @@ services:
...
     ports:
-      - 5433:5432
+      - ${POSTGRES_PORT:-5433}:5432
     user: postgres
     environment:
       POSTGRES_USER: algorand
       POSTGRES_PASSWORD: algorand
       POSTGRES_DB: indexer_db
+    volumes:
+      - ./volumes/indexer:/opt/mnt
```
Comment (tzaffi) [inline docker-compose.yml:66]:
```suggestion
```

File: sandbox [hunk 3]
```diff
@@ -1,6 +1,8 @@
 #!/usr/bin/env bash
 set -euo pipefail
 
+source .env
```
Comment (tzaffi) [inline sandbox:4]: allow overriding with `.env` variables for downstream users

File: images/algod/setup.py [hunk 4]
```diff
@@ -125,20 +135,35 @@ def configure_data_dir(network_dir, token, algod_port, kmd_port, bootstrap_url):
         startCommands = create_real_network(args.bin_dir, args.network_dir, args.network_template, args.genesis_file)
 
     # Write start script
-    print(f'Start commands for {args.start_script}:')
-    pp.pprint(startCommands)
+    print(f'ZZZZ: Start commands for {args.start_script}:')
+    pp.pprint(f'ZZZZ: startCommands={startCommands}')
```
Comment (tzaffi) [inline images/algod/setup.py:139]:
```suggestion
    print(f'Start commands for {args.start_script}:')
    pp.pprint(f'startCommands={startCommands}')
```

File: images/algod/setup.py [hunk 5]
```diff
@@ -102,11 +104,19 @@ def configure_data_dir(network_dir, token, algod_port, kmd_port, bootstrap_url):
...
     # Setup config, inject port
-    with open(join(node_dir, 'config.json'), 'w') as f:
-        f.write('{ "Version": 12, "GossipFanout": 1, "EndpointAddress": "0.0.0.0:%s", "DNSBootstrapID": "%s", "IncomingConnectionsLimit": 0, "Archival":false, "isIndexerActive":false, "EnableDeveloperAPI":true}' % (algod_port, bootstrap_url))
-    with open(join(kmd_dir, 'kmd_config.json'), 'w') as f:
-        f.write('{  "address":"0.0.0.0:%s",  "allowed_origins":["*"]}' % kmd_port)
-
+    node_config_path = join(node_dir, "config.json")
+    archival = 'true' if archival else 'false'
+    is_indexer_active = 'true' if is_indexer_active else 'false'
+    node_config = f'{{ "Version": 12, "GossipFanout": 1, "EndpointAddress": "0.0.0.0:{algod_port}", "DNSBootstrapID": "{bootstrap_url}", "IncomingConnectionsLimit": 0, "Archival":{archival}, "isIndexerActive":{is_indexer_active}, "EnableDeveloperAPI":true }}'
```
Comment (algochoi) [inline images/algod/setup.py:110]: nice f-str - is there a guarantee that `setup.py` will always run on python >=3.6?

File: sandbox [hunk 6]
```diff
@@ -203,7 +211,13 @@ sandbox () {
   #   dc_pty algod exec //bin/bash
   # This is to ensure compatibility with Windows Msys
   dc () {
-    $DOCKER_COMPOSE -f "$SANDBOX_DIR/docker-compose.yml" "$@"
+    VERBOSE_FLAG=""
+    if [ $VERBOSE_MODE == 1 ]; then
+      # --verbose is universally available since tag https://github.com/docker/compose-cli/releases/tag/v1.0.18 pr https://github.com/docker/compose-cli/pull/1830
```
Comment (algochoi) [inline sandbox:216]: Would this flag be passed into the `.env`? Would users be aware that this flag exists?

File: sandbox [hunk 7]
```diff
@@ -214,6 +228,7 @@ sandbox () {
   dc_pty () {
     # note: cannot be replaced by $WINDOWS_PTY_PREFIX dc "$@"
     #       because winpty requires an executable as argument
+    # TODO: can windows docker handle --verbose?
```
Comment (algochoi) [inline sandbox:231]: Are there existing windows build tests?

File: images/algod/setup.py (hunk 5 above)
Comment (tzaffi) [reply images/algod/setup.py:110]: We get to control that in Algod's Dockerfile. Though we're not explicitly ensuring python >= 3.6, I'm not the first to introduce f-strings here. (e.g. see [this PR from July 2020](https://github.com/algorand/sandbox/blob/d70ffd7c589b699cf275402ba187005097abdf02/images/algod/setup.py#L143))

File: docker-compose.yml [hunk 8]
```diff
@@ -34,7 +36,7 @@ services:
         BRANCH: "${INDEXER_BRANCH}"
         SHA: "${INDEXER_SHA}"
     ports:
-      - 8980:8980
+      - ${INDEXER_PORT:-8980}:8980
     restart: unless-stopped
     environment:
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply docker-compose.yml:41]: I believe not. `environment` is introducing env var's into the docker container. Indexer's docker container believes that it's serving on hard-coded port 8980 - as usual. `INDEXER_PORT` is just providing the translation from external port to internal port.

File: sandbox (hunk 7 above)
Comment (tzaffi) [reply sandbox:231]: No. Should we introduce?

@michaeldiamant @winder

File: sandbox (hunk 6 above)
Comment (tzaffi) [reply sandbox:216]: This isn't environmentally configurable. You can only control this via `./sandbox`'es arg `-v`

Comment (algochoi) [reply sandbox:216]: Oops, sounds good - resolving

File: README.md [hunk 9]
```diff
@@ -185,6 +185,8 @@ export NETWORK_TEMPLATE="images/algod/future_template.json"
 export NETWORK_NUM_ROUNDS=300000
 export NETWORK_BOOTSTRAP_URL=""
 export NETWORK_GENESIS_FILE=""
+export NODE_ARCHIVAL=""
+export NODE_V1_INDEXER=""
```
Comment (winder) [inline README.md:189]: Can we skip adding the v1 Indexer? It's been deprecated for years.

**Review by winder — CHANGES_REQUESTED** [review]
Other than the V1 indexer support, this looks great.

**Comment by tzaffi** [comment]
> Other than the V1 indexer support, this looks great.

The [latest commit](https://github.com/algorand/sandbox/pull/136/commits/24282301f17748b31fdd48425f27d717358226ff) remove indexer v1 support

**Review by winder — APPROVED** [review]

**Pull request merged**